### PR TITLE
Move TWS data from Route::Node to Route

### LINKED ISSUE #96: Refactor Node and Route structures
These structures are tied together quite closely, but that's not evident from how they are currently used. Further, these objects are clearly not in charge of their internal state: LS and the operators modify this repeatedly. 

Additionally, we do no really test these objects yet since they are mostly internal to the C++ implementation of LS. We should probably bind and test them in much the same way we did for `TimeWindowSegment`.

### PR DESCRIPTION
**Pull request by N-Wouda**
This PR:

- [x] Closes #96.
- [x] Adds and passes relevant tests.
- [x] Has well-formatted code and documentation.

### Benchmark on VRPTW

This PR (a9ba529e010d9614017fd3944e59bfc091f8c2ec; job ID 3382696):
```
                MEAN      MIN       MAX   COV
      inst
obj   C1     41357.5  41328.8   41382.7   0.0
      C2     16191.4  16186.3   16194.4   0.0
      R1     47287.8  47229.8   47325.3   0.1
      R2     27706.3  27692.4   27732.9   0.0
      RC1    44217.5  44201.7   44235.4   0.0
      RC2    23216.1  23204.4   23238.8   0.0
iters C1     73683.2  55804.6   80062.6   8.9
      C2    123546.9  95605.4  131344.4   8.4
      R1     40122.2  29845.2   44674.9  10.5
      R2     57048.2  45092.9   62249.0   8.2
      RC1    45799.1  33013.6   54637.5  11.8
      RC2    65087.7  48964.3   71194.2   9.8
```
This PR is some ten-ish percent faster than current main (from #320):
```
                mean       min       max
      inst
obj   C1     41359.9   41333.3   41383.4
      C2     16191.5   16186.3   16194.4
      R1     47307.9   47264.4   47350.1
      R2     27710.4   27696.9   27728.9
      RC1    44229.0   44207.2   44246.1
      RC2    23217.8   23205.0   23226.7
iters C1     67896.1   64152.1   72730.3
      C2    114640.7  110379.3  122336.3
      R1     35754.7   34111.1   37694.9
      R2     50979.6   46661.8   52468.4
      RC1    40433.6   38074.8   44908.3
      RC2    59870.8   55444.4   62651.8
```
I think the speed boost mostly comes from the better cache locality `Route::stats` provides. It might be interesting to also try making `Route::nodes` a `std::vector<Node>`, rather than `std::vector<Node *>`.

<details>

**Notes**:

- It is essential that you add a test when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by N-Wouda** [comment]
A stub; doesn't work yet. I'll finish this tomorrow, and that'll then close #96. There's more we can do, but I don't think that's super pressing right now.

**Comment by N-Wouda** [comment]
I'll run a benchmark.

**Comment by N-Wouda** [comment]
For my own edification, I'm also running a benchmark where we do not use the `twsBefore`/`twsAfter` caches, but instead recompute them every time they're needed:
```
RUNNING
```

<details>
Job ID: 3389178
</details>

**Comment by N-Wouda** [comment]
In another PR I'll bind `LocalSearchOperator::evaluate` and `LocalSearchOperator::apply`, so we can start writing really granular tests for the LS operators.

**Review by leonlan — APPROVED** [review]

File: pyvrp/cpp/search/Exchange.h [hunk 1]
```diff
@@ -87,58 +87,59 @@ Cost Exchange<N, M>::evalRelocateMove(Route::Node *U,
...
     {
-        if (U->route()->isFeasible() && deltaCost >= 0)
+        if (uRoute->isFeasible() && deltaCost >= 0)
             return deltaCost;
 
-        auto uTWS = TimeWindowSegment::merge(
-            data.durationMatrix(), p(U)->twBefore, n(endU)->twAfter);
+        auto uTWS = TimeWindowSegment::merge(data.durationMatrix(),
+                                             uRoute->twsBefore(U->idx() - 1),
+                                             uRoute->twsAfter(U->idx() + N));
```
[the remark this answers is not in the document]
Comment (N-Wouda) [reply pyvrp/cpp/search/Exchange.h:111]: > Question: why are the `p(U)` replaced with `U->idx() - 1`, is there a plan to get rid of `p()` and `n()` entirely?

Maybe. I think the compiler (currently, might change later if we move some stuff around again) struggles a bit to realise that the indices are all ordered, and thus `p(U)->idx() == U->idx() - 1`. Directly computing it is much easier for the thing. The disassembly suggests it's loading more with the old `p(U)` than it is with `U->idx() - 1 `, but I don't know about the immediate performance impact of it.

In any case, on the whole, this PR seems to be quite a bit faster than current main, so there's clearly something here that matters.

**Pull request merged**